### ISSUE TITLE
Direct3D11 texture memory leak

### ISSUE THREAD

**Issue by zerppa** [mellinoe/veldrid#192]
The following code demonstrates the leak. Debug from Visual Studio and observe how the process memory goes up constantly.

Reproducible with Veldrid 4.6.1, .NET Core 3, 64-bit Windows, and compiling with Any CPU. I have a fairly old GTX 660, but I don't think this is a driver issue.

Vulkan and OpenGL backends seem to work properly (Metal I haven't tested). Enforced garbage collection has no effect.

```csharp
namespace VeldridTest
{
    using Veldrid;
    using Veldrid.Sdl2;
    using Veldrid.StartupUtilities;
    using Veldrid.Utilities;

    class Program
    {
        static void Main(string[] args) => new Program().Run();

        void Run()
        {
            var windowCreateInfo = new WindowCreateInfo
            {
                X = Sdl2Native.SDL_WINDOWPOS_CENTERED,
                Y = Sdl2Native.SDL_WINDOWPOS_CENTERED,
                WindowWidth = 400,
                WindowHeight = 200,
                WindowTitle = "Test",
                WindowInitialState = WindowState.Normal
            };

            var window = VeldridStartup.CreateWindow(ref windowCreateInfo);
            window.Resizable = false;

            var options = new GraphicsDeviceOptions(
                debug: false,
                swapchainDepthFormat: null,
                syncToVerticalBlank: true,
                resourceBindingModel: ResourceBindingModel.Improved,
                preferDepthRangeZeroToOne: false,
                preferStandardClipSpaceYDirection: false);

            var graphicsDevice = VeldridStartup.CreateGraphicsDevice(window, options, GraphicsBackend.Direct3D11); // <-- Only Direct3D11 leaks
            var resourceFactory = new DisposeCollectorResourceFactory(graphicsDevice.ResourceFactory);
            var commandList = resourceFactory.CreateCommandList();
            Texture texture = null;
            TextureView textureView = null;

            while (window.Exists)
            {
                window.PumpEvents();
                if (window.Exists)
                {
                    commandList.Begin();
                    commandList.SetFramebuffer(graphicsDevice.MainSwapchain.Framebuffer);
                    commandList.ClearColorTarget(0, RgbaFloat.CornflowerBlue);

                    // Dispose resources we used last frame
                    if (texture != null)
                    {
                        texture.Dispose();
                        resourceFactory.DisposeCollector.Remove(texture);
                        textureView.Dispose();
                        resourceFactory.DisposeCollector.Remove(textureView);
                    }

                    // Allocate new resources
                    texture = resourceFactory.CreateTexture(TextureDescription.Texture2D(200, 30, 1, 1, PixelFormat.R8_G8_B8_A8_UNorm, TextureUsage.Sampled));
                    textureView = resourceFactory.CreateTextureView(texture);
                    // TODO: Also test with a ResourceSet using the TextureView

                    commandList.End();
                    graphicsDevice.SubmitCommands(commandList);
                    graphicsDevice.SwapBuffers(graphicsDevice.MainSwapchain);
                }
            }

            graphicsDevice.WaitForIdle();
            resourceFactory.DisposeCollector.DisposeAll();
            graphicsDevice.Dispose();
        }
    }
}
```

**Comment by mellinoe** [comment]
I do see a slow increase in memory usage for the first minute or so, but it's not a persistent growth, and it seems to stop growing entirely at some point (around 90-100 seconds on my machine). A native heap snapshot indicates that, initially, there is some memory being allocated by the Nvidia drivers, but only for those first 100ish seconds. There is no growth on the managed heap. On top of that, the D3D11 debug layers indicate that there are no D3D objects being leaked, so everything is being disposed properly. My guess is that the driver is being a little lax with its memory management or caching at first, but stops after a point. Can you confirm the same behavior on your machine, or are you seeing persistent growth which never stops?

**Comment by zerppa** [comment]
The memory hog is quite steep and reaches 1000MB at around 7:30. What's interesting though, is that most of it is due to _committed memory_. Both the _Working set_ and _Private bytes_ seem to grow a lot slower (although they do keep growing). For some reason the CLR keeps reserving a lot more of virtual memory than what's really needed.

So I decided to test this with another machine. With that I get similar results to what you're describing - memory usage grows steadily but at a much slower pace than with my first setup. I let it run for 10 minutes and after that the memory usage was around 200MB and still growing by 1MB every 4-5 seconds. Both Vulkan and OpenGL were completely stable.

Veldrid's code for disposing resources is pretty straight forward looking (and so is SharpDX). My original fear was that some COM pointers are being kept alive by other SharpDX resources. That might or might not be the case, but it seems that there's at least some unmanaged trash left behind.

All in all, Veldrid is probably not at fault here. It's either some SharpDX related stuff or a driver issue. Then again, a game probably should not be creating and disposing resources at such rate anyway. I found this when I was playing with text rendering where each string was first measured and then a texture was created for it.
